File: GildedRose/python/gilded_rose.py
from abc import ABC, abstractmethod


class Item:
    def __init__(self, name: str, sell_in: int, quality: int):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)


class ItemUpdater(ABC):

    @staticmethod
    @abstractmethod
    def update_item(item: Item):
        pass


def decrease_item_quality(item: Item, amount: int = 1) -> None:
    item.quality = max(item.quality - amount, 0)


def increase_item_quality(item: Item, amount: int = 1, max_quality: int = 50) -> None:
    item.quality = min(item.quality + amount, max_quality)


class DefaultItem(ItemUpdater):
    @staticmethod
    def _passed_sell_date(item: Item) -> bool:
        return item.sell_in <= 0

    @staticmethod
    def _update_quality(item: Item):
        decrease_item_quality(item)
        if DefaultItem._passed_sell_date(item):
            decrease_item_quality(item)

    @staticmethod
    def _update_sell_in(item: Item):
        item.sell_in = max(0, item.sell_in - 1)

    @staticmethod
    def update_item(item: Item):
        DefaultItem._update_quality(item)
        DefaultItem._update_sell_in(item)


class AgedBrie(ItemUpdater):

    @staticmethod
    def _update_quality(item: Item):
        increase_item_quality(item)

    @staticmethod
    def _update_sell_in(item: Item):
        item.sell_in = max(0, item.sell_in - 1)

    @staticmethod
    def update_item(item: Item):
        AgedBrie._update_quality(item)
        AgedBrie._update_sell_in(item)


class Sulfuras(ItemUpdater):
    @staticmethod
    def _update_quality(item: Item):
        item.quality = min(item.quality, 80)

    @staticmethod
    def _update_sell_in(item: Item):
        item.sell_in = max(0, item.sell_in)

    @staticmethod
    def update_item(item: Item):
        Sulfuras._update_sell_in(item)
        Sulfuras._update_quality(item)


class BackstagePasses(ItemUpdater):
    @staticmethod
    def _update_quality(item: Item):
        increase_item_quality(item)

        if item.sell_in <= 10:
            increase_item_quality(item)

        if item.sell_in <= 5:
            increase_item_quality(item)

        if item.sell_in <= 0:
            item.quality = 0

    @staticmethod
    def _update_sell_in(item: Item):
        item.sell_in = max(0, item.sell_in - 1)

    @staticmethod
    def update_item(item: Item):
        BackstagePasses._update_quality(item)
        BackstagePasses._update_sell_in(item)


class Conjured(ItemUpdater):
    @staticmethod
    def _update_quality(item: Item):
        decrease_item_quality(item, 2)

    @staticmethod
    def _update_sell_in(item: Item):
        item.sell_in = max(0, item.sell_in - 1)

    @staticmethod
    def update_item(item: Item):
        Conjured._update_quality(item)
        Conjured._update_sell_in(item)


item_mapping: dict[str, ItemUpdater] = {
    "Aged Brie": AgedBrie,
    "Sulfuras, Hand of Ragnaros": Sulfuras,
    "Backstage passes to a TAFKAL80ETC concert": BackstagePasses,
    "Conjured": Conjured,
}


class GildedRose:

    def __init__(self, items: list[Item]):
        self.items = items

    def update_quality(self):
        for item in self.items:
            updater = item_mapping.get(item.name, DefaultItem)
            updater.update_item(item)

File: GildedRose/go/gildedrose/gildedrose.go
package gildedrose

type Item struct {
	Name    string
	SellIn  int8
	Quality int8
}

const DEFAULT_MAX_QUALITY = 50
const DEFAULT_MIN_QUALITY = 0
const SULFURAS_MAX_QUALITY = 80

func decreaseItemQuality(item *Item, minValue int8) {
	if item.Quality > minValue {
		item.Quality = item.Quality - 1
	}
}
func increaseItemQuality(item *Item, maxValue int8) {
	if item.Quality < maxValue {
		item.Quality = item.Quality + 1
	}
}

type ItemUpdater interface {
	uptadeItem(item *Item)
}

type DefaultUpdater struct {
}

func (DefaultUpdater) uptadeItem(item *Item) {
	item.SellIn = item.SellIn - 1
	decreaseItemQuality(item, DEFAULT_MIN_QUALITY)
	if item.SellIn < 0 {
		decreaseItemQuality(item, DEFAULT_MIN_QUALITY)
	}
}

type AgedBrieUpdater struct {
}

func (AgedBrieUpdater) uptadeItem(item *Item) {
	item.SellIn = item.SellIn - 1
	increaseItemQuality(item, DEFAULT_MAX_QUALITY)
}

type SulfurasUpdater struct {
}

func (SulfurasUpdater) uptadeItem(item *Item) {
	if item.Quality > SULFURAS_MAX_QUALITY {
		item.Quality = SULFURAS_MAX_QUALITY
	}
}

type BackstagePassesUpdater struct {
}

func (BackstagePassesUpdater) uptadeItem(item *Item) {
	item.SellIn = item.SellIn - 1
	increaseItemQuality(item, DEFAULT_MAX_QUALITY)
	if item.SellIn < 10 {
		increaseItemQuality(item, DEFAULT_MAX_QUALITY)
	}
	if item.SellIn < 5 {
		increaseItemQuality(item, DEFAULT_MAX_QUALITY)
	}
	if item.SellIn < 0 {
		item.Quality = 0
	}
}

type ConjuredUpdater struct {
}

func (ConjuredUpdater) uptadeItem(item *Item) {
	item.SellIn = item.SellIn - 1
	decreaseItemQuality(item, DEFAULT_MIN_QUALITY)
	decreaseItemQuality(item, DEFAULT_MIN_QUALITY)
	if item.SellIn < 0 {
		decreaseItemQuality(item, DEFAULT_MIN_QUALITY)
		decreaseItemQuality(item, DEFAULT_MIN_QUALITY)
	}
}

type UpdaterFactory struct {
	updaterMap map[string]ItemUpdater
}

func NewUpdaterFactory() UpdaterFactory {
	updaterMap := make(map[string]ItemUpdater)
	updaterMap["Aged Brie"] = AgedBrieUpdater{}
	updaterMap["Sulfuras, Hand of Ragnaros"] = SulfurasUpdater{}
	updaterMap["Backstage passes to a TAFKAL80ETC concert"] = BackstagePassesUpdater{}
	updaterMap["Conjured"] = ConjuredUpdater{}
	return UpdaterFactory{updaterMap}
}

func (factory *UpdaterFactory) get(item *Item) ItemUpdater {
	updater, ok := factory.updaterMap[item.Name]
	if ok {
		return updater
	}
	return DefaultUpdater{}
}

func UpdateQuality(items []*Item) {
	factory := NewUpdaterFactory()
	for _, item := range items {
		updater := factory.get(item)
		updater.uptadeItem(item)
	}
}
